backend/vkt/src/main/java/fi/oph/vkt/service/PaymentService.java

```diff
@@ -0,0 +1,212 @@
+    }
+
+    return enrollment;
+  }
+
+  private Payment finalizePayment(final Long paymentId, final Map<String, String> paymentParams)
+    throws IOException, InterruptedException {
+    final Payment payment = paymentRepository
+      .findById(paymentId)
+      .orElseThrow(() -> new NotFoundException("Payment not found"));
+    final PaymentStatus currentStatus = payment.getPaymentStatus();
+    final PaymentStatus paymentStatus = PaymentStatus.fromString(paymentParams.get("checkout-status"));
+
+    if (!paytrailService.validate(paymentParams)) {
+      LOG.error("Payment ({}) validation failed for params {}", paymentId, paymentParams);
+      throw new APIException(APIExceptionType.PAYMENT_VALIDATION_FAIL);
+    }
+
+    if (currentStatus == PaymentStatus.OK && paymentStatus != PaymentStatus.OK) {
+      LOG.error("Cannot change payment status from OK to {} for payment {}", paymentStatus, paymentId);
+      throw new APIException(APIExceptionType.PAYMENT_ALREADY_PAID);
+    }
+
+    // Already paid. Payment url may be called multiple times
+    if (currentStatus == PaymentStatus.OK) {
+      return payment;
+    }
+
+    final int amount = Integer.parseInt(paymentParams.get("checkout-amount"));
+    if (amount != payment.getAmount()) {
+      LOG.error("Payment ({}) amount ({}) does not match expected amount ({})", paymentId, amount, payment.getAmount());
+      throw new APIException(APIExceptionType.PAYMENT_AMOUNT_MISMATCH);
+    }
+
+    final Enrollment enrollment = updateEnrollmentStatus(payment, paymentStatus);
+    payment.setPaymentStatus(paymentStatus);
+
+    enrollmentRepository.saveAndFlush(enrollment);
+    paymentRepository.saveAndFlush(payment);
+
+    final Person person = payment.getPerson();
+    if (paymentStatus == PaymentStatus.OK) {
+      publicEnrollmentEmailService.sendEnrollmentConfirmationEmail(enrollment, person);
+    }
+
+    return payment;
+  }
+
+  @Transactional
+  public String success(final Long paymentId, final Map<String, String> paymentParams)
+    throws IOException, InterruptedException {
+    final String baseUrl = environment.getRequiredProperty("app.base-url.public");
+    final Payment payment = finalizePayment(paymentId, paymentParams);
+    final ExamEvent examEvent = findExam(payment);
+
+    return String.format("%s/ilmoittaudu/%d/maksu/valmis", baseUrl, examEvent.getId());
+  }
+
+  @Transactional
+  public String cancel(final Long paymentId, final Map<String, String> paymentParams)
+    throws IOException, InterruptedException {
+    final String baseUrl = environment.getRequiredProperty("app.base-url.public");
+    final Payment payment = finalizePayment(paymentId, paymentParams);
+    final ExamEvent examEvent = findExam(payment);
+
+    return String.format("%s/ilmoittaudu/%d/maksu/peruutettu", baseUrl, examEvent.getId());
+  }
+
+  @Transactional
+  public String create(final Long enrollmentId, final Person person) {
```

Comment:
Nimeksi vois ehkä vaihtaa createPaymentForEnrollment tms. Tai sit jättää vaan näin kun ei meillä muitakaan maksuja ole.